### PR TITLE
Replaced environment variables with parameters from AWS parameter store

### LINKED ISSUE #44: Use AWS parameter store
We would like to start using the [`parameter store`](https://eu-west-1.console.aws.amazon.com/systems-manager/parameters/?region=eu-west-1&tab=Table) from AWS to store the settings from our project.

What is AWS parameter store?
You can read about it on google. Here are 2 links that I found useful [link 1](https://medium.com/@nqbao/how-to-use-aws-ssm-parameter-store-easily-in-python-94fda04fea84) | [link 2](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html)

Why use it?
Mostly because it allows us to change configuration at any time and the application will read the new configuration when a new spider is executed. This means that we won't have to restart the containers every time we want to make some config changes.

What we want to change exactly?
We want to start using the parameter store for `config.py` file. Currently we are using `os.environ.get('param name')` (they are [environment variables](https://en.wikipedia.org/wiki/Environment_variable)) to pass the configuration to the project. We would like to replace this logic and instead use the `parameter store`.
We will be storing all the parameters in AWS and we will need to implement the code that will get those settings and pass them to the `config.py` file.

### PR DESCRIPTION
**Pull request by VladAlex98**
Fix #44

### CODE REVIEW

File: ScrapyKeeper/config.py [hunk 1]
```diff
@@ -1,14 +1,34 @@
 # Statement for enabling the development environment
+import logging
 import os
+import boto3
+from botocore.exceptions import ClientError
 
 DEBUG = True
 
+
+def get_param_from_client(client, parameter_name):
```
Comment (c-lorand) [inline ScrapyKeeper/config.py:10]: You should create a class and initialize the client in the constructor. After that use that client and don't send it to this function every time you call it.

File: ScrapyKeeper/config.py [hunk 2]
```diff
@@ -1,14 +1,34 @@
...
+    try:
+        value = client.get_parameter(Name=parameter_name, WithDecryption=True)['Parameter']['Value']
+    except ClientError:
+        logging.error(' ClientError: There is no parameter named: {}'.format(parameter_name))
+        return
```
Comment (c-lorand) [inline ScrapyKeeper/config.py:15]: in order to get rid of this `return` statement that's sitting on it's own here, you could move everything inside the `try` block.

File: ScrapyKeeper/config.py [hunk 3]
```diff
@@ -1,14 +1,34 @@
...
+
+
 # Define the application directory
 
 
 BASE_DIR = os.path.abspath(os.path.dirname(__file__))
+ssm_client = boto3.client('ssm', region_name='eu-west-1')
 
-SQLALCHEMY_DATABASE_URI = os.getenv('DATABASE_CONNECTION_STRING', None)
+SQLALCHEMY_DATABASE_URI = get_param_from_client(ssm_client, '/data/crawlie-keeper/DATABASE_CONNEaCTION_STRING') or None
```
Comment (c-lorand) [inline ScrapyKeeper/config.py:31]:
```suggestion
SQLALCHEMY_DATABASE_URI = get_param_from_client(ssm_client, '/data/crawlie-keeper/DATABASE_CONNECTION_STRING') or None
```

File: ScrapyKeeper/config.py [hunk 4]
```diff
@@ -47,24 +67,24 @@
 BASIC_AUTH_FORCE = True
 
 NO_SENTRY = False
-SENTRY_URI = os.getenv('SENTRY_URI', None)
+SENTRY_URI = get_param_from_client(ssm_client, '/data/crawlie-keeper/SENTRY_URI') or None
```
Comment (c-lorand) [inline ScrapyKeeper/config.py:70]: This looks like it returns `None` by feault if there is no parameter in the store. The `or` block is not needed anymore.

File: ScrapyKeeper/ParameterStore.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import boto3
```
Comment (c-lorand) [inline ScrapyKeeper/ParameterStore.py:1]: Move this file into `util` folder and change the name to be lowercase to keep consistency.

File: ScrapyKeeper/ParameterStore.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+            if value == 'None':
+                return None
+            elif value == '\'\'':
+                return ''
+
+            return value
+
+        except ClientError:
+            return default
```
Comment (c-lorand) [inline ScrapyKeeper/ParameterStore.py:23]: We should find a way of returning the default value only when it's sent by the user and throw an error when both the param and the default are missing.
Maybe try `kwargs`?

File: ScrapyKeeper/config.py [hunk 7]
```diff
@@ -1,14 +1,16 @@
...
 import os
+from ScrapyKeeper.ParameterStore import ParameterStore
 
 DEBUG = True
 
 # Define the application directory
 
 
 BASE_DIR = os.path.abspath(os.path.dirname(__file__))
+ssm_client = ParameterStore()
```
Comment (c-lorand) [inline ScrapyKeeper/config.py:11]: Rename this variable to be more descriptive.
```suggestion
param_store = ParameterStore()
```

File: ScrapyKeeper/app/util/parameter_store.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+                return None
+            elif value == '\'\'':
+                return ''
+
+            return value
+
+        except ClientError:
+            if not default_value:
+                logging.error(' ClientError: Missing parameter: "{}" and default value'.format(parameter_name))
+                return
```
Comment (c-lorand) [inline ScrapyKeeper/app/util/parameter_store.py:24]: Maybe you could raise an exception here.
Just an idea, it's ok as it is as well.

**Comment by georgetutuianu** [comment]
The project should run locally without connecting to AWS. Are we gonna be able to do it once this PR is merged?
@VladAlex98 @c-lorand ?

**Comment by VladAlex98** [comment]
> The project should run locally without connecting to AWS. Are we gonna be able to do it once this PR is merged?
> @VladAlex98 @c-lorand ?

The project is not working without the credentials. That means you must connect at least once to AWS, or we can change the `ParameterStore` class to raise the `NoCredentialsError` and return the default_value instead of getting the error.

File: ScrapyKeeper/config.py [hunk 9]
```diff
@@ -1,14 +1,17 @@
...
 DEBUG = True
 
 # Define the application directory
 
 
 BASE_DIR = os.path.abspath(os.path.dirname(__file__))
+param_store = ParameterStore()
 
-SQLALCHEMY_DATABASE_URI = os.getenv('DATABASE_CONNECTION_STRING', None)
+SQLALCHEMY_DATABASE_URI = param_store.get_param('/data/crawlie-keeper/DATABASE_CONNECTION_STRING')
```
Comment (c-lorand) [inline ScrapyKeeper/config.py:14]: get the default value from env please

File: ScrapyKeeper/config.py [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
 DEBUG = True
 
 # Define the application directory
 
 
 BASE_DIR = os.path.abspath(os.path.dirname(__file__))
+param_store = ParameterStore()
 
-SQLALCHEMY_DATABASE_URI = os.getenv('DATABASE_CONNECTION_STRING', None)
+SQLALCHEMY_DATABASE_URI = param_store.get_param('/data/crawlie-keeper/DATABASE_CONNECTION_STRING', None)
```
Comment (c-lorand) [inline ScrapyKeeper/config.py:14]: I insist that you should get the default value from `env`
`os.getenv('DATABASE_CONNECTION_STRING', None)`

**Pull request merged**